src/Formats/ProtobufWriter.cpp

```diff
@@ -123,21 +124,38 @@ namespace
 
 // SimpleWriter is an utility class to serialize protobufs.
 // Knows nothing about protobuf schemas, just provides useful functions to serialize data.
-ProtobufWriter::SimpleWriter::SimpleWriter(WriteBuffer & out_) : out(out_), current_piece_start(0), num_bytes_skipped(0)
+ProtobufWriter::SimpleWriter::SimpleWriter(WriteBuffer & out_, const bool single_message_mode_)
+    : out(out_)
+    , current_piece_start(0)
+    , num_bytes_skipped(0)
+    , produce_length_delimiters(!single_message_mode_)
+    // normally for single_message_mode we forbid outputting more than one row
+    // (it would lead to malformed protobuf message), but Kafka/Rabbit push every row
+    // in a separate message, so it's valid case there and should be allowed.
+    , allow_several_messages(!single_message_mode_ || out.producesIsolatedRows())
```

Comment:
If they push every row in a separate message, the second `startMessage` call won't arrive? Can we do away with the `allow_several_messages` then, and check for something like __this is the first message, or we do produce length delimiters__? Maybe we don't even need `row_was_send` and can use the number of bytes written instead.

Reply:
> If they push every row in a separate message, the second startMessage call won't arrive? 

In case when we need to format > 1 row, for normal buffer we need to send exception after the first row, for Kafka buffer we can send any number of rows w/o delimiters (as that Kafka buffer will split them in separate messages). 

startMessage called before every new row, see https://github.com/ClickHouse/ClickHouse/blob/2dba96717545a22e2bcb380772d4e6c26c21c107/src/Processors/Formats/Impl/ProtobufRowOutputFormat.cpp#L35-L41 

If we want to optimize out 2 bytes and couple of cpu ticks, it can be rewritten using 2 vars, instead of 3. 

For example, it can be 
```
const delimiters  enum( needed = 0, no_delimiters /* for Kafka */, no_delimiters_single_row /* the rest */);
 + smth like row_was_send 
```
or smth like FSM on single var
```
 state enum (producing_delimiters = 0, producing_no_delimiters =1, producing_first_nondelimited_row=2, first_nondelimited_row_produced=3)
```
Any preferences / better options? 

> Maybe we don't even need row_was_send and can use the number of bytes written instead.

number of bytes written ? SimpleWriter doesn't have such properly...